locator/tablets.hh

```diff
@@ -0,0 +1,177 @@
+
+// Identifies tablet within the scope of a single tablet_map,
+// which has a scope of (table_id, token metadata version).
+// Different tablets of different tables can have the same tablet_id.
+// Different tablets in subsequent token metadata version can have the same tablet_id.
+// When splitting a tablet, one of the new tablets (in the new token metadata version)
+// will have the same tablet_id as the old one.
+using tablet_id = size_t;
+
+struct tablet_replica {
+    host_id host;
+    shard_id shard;
+
+    bool operator==(const tablet_replica&) const = default;
+};
+
+std::ostream& operator<<(std::ostream&, const tablet_replica&);
+
+using tablet_replica_set = utils::small_vector<tablet_replica, 3>;
+
+/// Stores information about a single tablet.
+struct tablet_info {
+    tablet_replica_set replicas;
+
+    std::optional<shard_id> get_shard(host_id host) const {
+        for (auto&& r : replicas) {
+            if (r.host == host) {
+                return r.shard;
+            }
+        }
+        return std::nullopt;
+    }
+
+    bool operator==(const tablet_info&) const = default;
+};
+
+/// Used for storing tablet state transition during topology changes.
+/// Describes transition of a single tablet.
+struct tablet_transition_info {
+    tablet_replica_set next;
+    tablet_replica pending_replica; // Optimization (next - tablet_info::replicas)
+
+    bool operator==(const tablet_transition_info&) const = default;
+};
+
+/// Stores information about tablets of a single table.
+class tablet_map {
+public:
+    using tablet_container = utils::chunked_vector<tablet_info>;
+private:
+    // The implementation assumes that _tablets.size() is a power of 2:
+    //
+    //   _tablets.size() == 1 << _log2_tablets
+    //
+    tablet_container _tablets;
+    size_t _log2_tablets; // log_2(_tablets.size())
+    std::unordered_map<tablet_id, tablet_transition_info> _transitions;
+public:
+    tablet_map();
+
+    /// Constructs a tablet map.
+    /// The actual number of allocated tablets may be higher due to rounding up to the power of two.
+    /// Check tablet_count().
+    ///
+    /// \param tablet_count The desired tablets to allocate.
+    explicit tablet_map(size_t tablet_count);
+
+    tablet_id get_tablet_id(token) const;
+    const tablet_info& get_tablet_info(tablet_id) const;
+    const tablet_transition_info* get_tablet_transition_info(tablet_id) const;
+    dht::token get_last_token(tablet_id id) const;
+    dht::token get_first_token(tablet_id id) const;
+    dht::token_range get_token_range(tablet_id id) const;
+
+    std::optional<tablet_id> next_tablet(tablet_id t) const {
+        if (t == tablet_count()) {
+            return std::nullopt;
+        }
+        return t + 1;
+    }
+
+    const tablet_container& tablets() const {
+        return _tablets;
+    }
+
+    size_t tablet_count() const {
+        return _tablets.size();
+    }
+
+    const tablet_info& get_tablet_info(token t) const {
```

Comment:
nit: noexcept
(`get_tablet_id(t)` must return a valid tablet_id for any `token t`, no?)

Reply:
Yes, we always have a mapping. But why noexcept? It's limiting to the implementation.

Reply:
otherwise, I'd expect error handling by the callers of this function.
If it doesn't throw at the moment, by design, marking it as noexcept will simplify the life of the callers.
If, in the future, we'd like to throw from this function, we then need to survey all call paths to make sure they'd handle such exception properly.

Reply:
If the caller needs noexcept in the future, it's easy to slap noexcept because it's easy to prove this is noexcept. If we did this preemptively if there is no need, and the implementation wants to throw, we would have to prove that users don't depend on noexcept, which in general is much harder (there are many of them, and it's not obvious if they rely on it). It is also a risk to mark it as noexcept, because that's an extra constraint which we can miss to fulfill and cause abort.

Reply:
@avikivity What's your take on this?

Reply:
Here, since all tokens must be mapped to tablets I think it should be noexcept by design.
If in the future we'd need to e.g. allocate tablet_info on-demand, I think we should expose that as `get_or_create_tablet_info` and let this function assume it is already allocated and if not terminate on internal error.

Reply:
I'm not a fan of strict noexcept.

It has its place in move constructors and assignment operators, and in functions that return a future. But in random functions it's easy to add noexcept on Monday and change the implementation to something that throws on Thursday. Or even on the same day in the same patch.

Nothing original about this, I've seen the argument from @tgrabiec too.